src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -699,48 +697,128 @@ public SolutionCompilationState RemoveAnalyzerReference(ProjectId projectId, Ana
             texts,
             arg: mode,
             updateDocument: static (oldDocumentState, text, mode) =>
-                SourceTextIsUnchanged(oldDocumentState, text) ? oldDocumentState : oldDocumentState.UpdateText(text, mode));
+                SourceTextIsUnchanged(oldDocumentState, text) ? oldDocumentState : oldDocumentState.UpdateText(text, mode),
+            contentChanged: true);
 
     private static bool SourceTextIsUnchanged(DocumentState oldDocument, SourceText text)
         => oldDocument.TryGetText(out var oldText) && text == oldText;
 
-    private SolutionCompilationState UpdateDocumentsInMultipleProjects<TDocumentData, TArg>(
-        ImmutableArray<(DocumentId documentId, TDocumentData documentData)> updates,
+    /// <summary>
+    /// Applies an update operation <paramref name="updateDocument"/> to specified <paramref name="documentsToUpdate"/>.
+    /// Documents may be in different projects.
+    /// </summary>
+    /// <param name="contentChanged">
+    /// True if <paramref name="updateDocument"/> changes the content of the document
+    /// (i.e. document text, not just document attributes).
+    /// </param>
+    private SolutionCompilationState UpdateDocumentsInMultipleProjects<TDocumentState, TDocumentData, TArg>(
+        ImmutableArray<(DocumentId documentId, TDocumentData documentData)> documentsToUpdate,
         TArg arg,
-        Func<DocumentState, TDocumentData, TArg, DocumentState> updateDocument)
+        Func<TDocumentState, TDocumentData, TArg, TDocumentState> updateDocument,
+        bool contentChanged)
+        where TDocumentState : TextDocumentState
+    {
+        return WithDocumentStatesOfMultipleProjects(
+            documentsToUpdate
+                .GroupBy(static d => d.documentId.ProjectId)
+                .Select(g =>
+                {
+                    var projectId = g.Key;
+                    var oldProjectState = SolutionState.GetRequiredProjectState(projectId);
+                    var oldDocumentStates = oldProjectState.GetDocumentStates<TDocumentState>();
+
+                    using var _ = ArrayBuilder<TDocumentState>.GetInstance(out var newDocumentStates);
+                    foreach (var (documentId, documentData) in g)
+                    {
+                        var oldDocumentState = oldDocumentStates.GetRequiredState(documentId);
+                        var newDocumentState = updateDocument(oldDocumentState, documentData, arg);
+
+                        if (ReferenceEquals(oldDocumentState, newDocumentState))
+                            continue;
+
+                        newDocumentStates.Add(newDocumentState);
+                    }
+
+                    return (projectId, newDocumentStates.ToImmutableAndClear());
+                }),
+            (oldProjectState, newDocumentStates) => GetUpdateDocumentsTranslationAction(oldProjectState, newDocumentStates, contentChanged));
```

Comment:
>(oldProjectState, newDocumentStates) => GetUpdateDocumentsTranslationAction(oldProjectState, newDocumentStates, contentChanged)); [](http://example.com/codeflow?start=12&length=129)

Should WithDocumentStatesOfMultipleProjects take in an arg to avoid this closure?

Reply:
Not sure it's worth the extra noise. We already allocate enumerators in `GroupBy` and `Select` here. One more allocation for the closure shouldn't make much difference.

Reply:
Up to you, but this is an area that definitely shows up in allocation profiles. There's a reason almost all code in this file uses static lambdas.